File: modules/napplet-host/android/src/main/java/org/nostrocket/hypergolic/host/PublishedRelayWebSocketHandshake.java
package org.nostrocket.hypergolic.host;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

/** Strict RFC 6455 server-upgrade response validator. Input must contain headers only. */
public final class PublishedRelayWebSocketHandshake {
  public static final int MAX_HEADER_BYTES = 16_384;
  private static final String GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";

  private PublishedRelayWebSocketHandshake() {}

  public static void validate(byte[] response, byte[] nonce) {
    if (response == null || nonce == null || nonce.length != 16) fail();
    if (response.length == 0 || response.length > MAX_HEADER_BYTES) fail();
    String raw = ascii(response);
    int end = raw.indexOf("\r\n\r\n");
    if (end < 0 || end + 4 != raw.length()) fail(); // Reject bodies and prefetched frame bytes.
    String[] lines = raw.substring(0, end).split("\r\n", -1);
    if (lines.length < 2 || !lines[0].matches("HTTP/1\\.1 101(?: [\\x20-\\x7e]*)?")) fail();
    Set<String> seen = new HashSet<>();
    String upgrade = null, connection = null, accept = null;
    for (int i = 1; i < lines.length; i++) {
      String line = lines[i];
      int colon = line.indexOf(':');
      if (colon <= 0 || line.charAt(0) == ' ' || line.charAt(0) == '\t') fail();
      String name = line.substring(0, colon);
      if (!name.matches("[!#$%&'*+.^_`|~0-9A-Za-z-]+")) fail();
      String key = name.toLowerCase(Locale.ROOT);
      if (!seen.add(key)) fail();
      String value = trimOws(line.substring(colon + 1));
      if (hasForbiddenControls(value)) fail();
      switch (key) {
        case "upgrade": upgrade = value; break;
        case "connection": connection = value; break;
        case "sec-websocket-accept": accept = value; break;
        case "sec-websocket-extensions": case "sec-websocket-protocol": case "location":
        case "content-length": case "transfer-encoding": fail(); break;
        default: break;
      }
    }
    if (!hasToken(upgrade, "websocket") || !hasToken(connection, "upgrade")) fail();
    String key = Base64.getEncoder().encodeToString(nonce) + GUID;
    String expected;
    try {
      expected = Base64.getEncoder().encodeToString(
          MessageDigest.getInstance("SHA-1").digest(key.getBytes(StandardCharsets.US_ASCII)));
    } catch (NoSuchAlgorithmException impossible) { throw new AssertionError(impossible); }
    if (accept == null || !MessageDigest.isEqual(
        expected.getBytes(StandardCharsets.US_ASCII), accept.getBytes(StandardCharsets.US_ASCII))) fail();
  }

  private static String ascii(byte[] bytes) {
    StringBuilder out = new StringBuilder(bytes.length);
    for (byte b : bytes) {
      int c = b & 0xff;
      if (c < 0x20 && c != '\r' && c != '\n' && c != '\t' || c > 0x7e) fail();
      out.append((char) c);
    }
    return out.toString();
  }
  private static boolean hasForbiddenControls(String value) {
    for (int i = 0; i < value.length(); i++) {
      char c = value.charAt(i);
      if (c < 0x20 && c != '\t' || c == 0x7f) return true;
    }
    return false;
  }
  private static String trimOws(String value) {
    int start = 0, end = value.length();
    while (start < end && (value.charAt(start) == ' ' || value.charAt(start) == '\t')) start++;
    while (end > start && (value.charAt(end - 1) == ' ' || value.charAt(end - 1) == '\t')) end--;
    return value.substring(start, end);
  }
  private static boolean hasToken(String value, String token) {
    if (value == null) return false;
    for (String part : value.split(",", -1)) if (trimOws(part).equalsIgnoreCase(token)) return true;
    return false;
  }
  private static void fail() { throw new IllegalArgumentException("Invalid WebSocket upgrade response"); }
}

File: modules/napplet-host/ios/PublishedRelayWebSocketHandshake.swift
import CryptoKit
import Foundation

/// Strict RFC 6455 server-upgrade response validator. Input must contain headers only.
enum PublishedRelayWebSocketHandshake {
    static let maxHeaderBytes = 16_384
    private static let guid = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

    static func validate(_ response: Data, nonce: Data) throws {
        guard nonce.count == 16, !response.isEmpty, response.count <= maxHeaderBytes,
              let raw = String(bytes: response, encoding: .ascii),
              let marker = raw.range(of: "\r\n\r\n"), marker.upperBound == raw.endIndex else {
            throw invalid()
        }
        let lines = raw[..<marker.lowerBound].components(separatedBy: "\r\n")
        guard lines.count >= 2, validStatus(lines[0]) else { throw invalid() }
        var seen = Set<String>()
        var upgrade: String?
        var connection: String?
        var accept: String?
        for line in lines.dropFirst() {
            guard let colon = line.firstIndex(of: ":"), colon != line.startIndex else { throw invalid() }
            let name = String(line[..<colon])
            guard name.utf8.allSatisfy(isTokenByte) else { throw invalid() }
            let key = name.lowercased()
            guard seen.insert(key).inserted else { throw invalid() }
            let value = trimOWS(String(line[line.index(after: colon)...]))
            guard !value.utf8.contains(where: isForbiddenControl) else { throw invalid() }
            switch key {
            case "upgrade": upgrade = value
            case "connection": connection = value
            case "sec-websocket-accept": accept = value
            case "sec-websocket-extensions", "sec-websocket-protocol", "location",
                 "content-length", "transfer-encoding": throw invalid()
            default: break
            }
        }
        guard hasToken(upgrade, "websocket"), hasToken(connection, "upgrade"),
              let accept,
              constantTimeEqual(Array(accept.utf8), Array(expectedAccept(nonce).utf8)) else {
            throw invalid()
        }
    }

    private static func validStatus(_ value: String) -> Bool {
        guard value.hasPrefix("HTTP/1.1 101") else { return false }
        let suffix = value.dropFirst("HTTP/1.1 101".count)
        return suffix.isEmpty || (suffix.first == " " && suffix.utf8.allSatisfy { $0 >= 0x20 && $0 <= 0x7e })
    }
    private static func expectedAccept(_ nonce: Data) -> String {
        var source = Data(nonce.base64EncodedString().utf8)
        source.append(contentsOf: guid.utf8)
        return Data(Insecure.SHA1.hash(data: source)).base64EncodedString()
    }
    private static func hasToken(_ value: String?, _ token: String) -> Bool {
        value?.split(separator: ",", omittingEmptySubsequences: false)
            .contains { trimOWS(String($0)).caseInsensitiveCompare(token) == .orderedSame } ?? false
    }
    private static func trimOWS(_ value: String) -> String {
        var start = value.startIndex, end = value.endIndex
        while start < end && (value[start] == " " || value[start] == "\t") { start = value.index(after: start) }
        while end > start {
            let prior = value.index(before: end)
            guard value[prior] == " " || value[prior] == "\t" else { break }
            end = prior
        }
        return String(value[start..<end])
    }
    private static func isTokenByte(_ byte: UInt8) -> Bool {
        (byte >= 48 && byte <= 57) || (byte >= 65 && byte <= 90) || (byte >= 97 && byte <= 122)
            || Array("!#$%&'*+-.^_|~".utf8).contains(byte) || byte == 96
    }
    private static func isForbiddenControl(_ byte: UInt8) -> Bool {
        (byte < 0x20 && byte != 0x09) || byte == 0x7f
    }
    private static func constantTimeEqual(_ lhs: [UInt8], _ rhs: [UInt8]) -> Bool {
        guard lhs.count == rhs.count else { return false }
        return zip(lhs, rhs).reduce(UInt8(0)) { $0 | ($1.0 ^ $1.1) } == 0
    }
    private static func invalid() -> NSError {
        NSError(domain: "PublishedRelayWebSocketHandshake", code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Invalid WebSocket upgrade response"])
    }
}
